### ISSUE TITLE
a question about auto grad. thx

### ISSUE THREAD

**Issue by yebangyu** [EdwardRaff/Inside-Deep-Learning#11]
Dear Edward,

From page 21 to page 23, when we are talking about auto grad, 

we choose to test  conditon ||prev - cur || < epsilon satisfies or not to check whether we have got the minimun

my question is : why not just to test whether the grad of cur is zero or not ?

that is to say : 

can

`while torch.linalg.norm(x_cur-x_prev) > epsilon:`

be replaced by

`epsilon = 1e-12 # an enough small value`

`while abs(cur.grad) > epsilon:`

?

thanks a lot !

**Comment by EdwardRaff** [comment]
Good question! For most neural networks we don’t use either of these kinds of stopping conditions. So it’s more an example of optimization than ideal way to train a neural network.The broader answer in the context of optimization problems is that both are valid with different trade offs. The zero check gets you closer to a local minima The difference between previous and current might stop before you reach a local minima, but it also might save you from waiting a very long time if the function is badly behaved and converging very slowly. I also just like the second approach because it’s more general purpose. Sometimes you want to check the convergence of something other than the gradient, which may not converge/minimize near zero. Sent from my iPhoneOn May 20, 2023, at 3:27 PM, yebangyu ***@***.***> wrote:﻿
Dear Edward,
From page 21 to page 23, when we are talking about auto grad,
we choose to test  conditon ||prev - cur || < epsilon satisfies or not to check whether we have got the minimun
my question is : why not just to test whether the grad of cur is zero or not ?
that is to say :
can
while torch.linalg.norm(x_cur-x_prev) > epsilon:
be replaced by
epsilon = 1e-12 # an enough small value while abs(cur.grad) > epsilon: 
?
thanks a lot !

**Comment by yebangyu** [comment]
thanks for your reply,  Edward

After a bit of thinking , I just consider that both of these approaches seem equivalent.

in sgd we have:

x_cur = x_prev  - learning_rate * grad

when grad->0 then learning_rate * grad will also be close to zero  so we have x_cur -> x_prev (x_cur is approximately equal to x_prev)

Am i right ?


[词典]	[is approximately equal to](https://www.baidu.com/link?url=CWKmuCuLchtVf9CRS3G0CiXdNw9vIz_j7Y9p-B0b1o2AKRU8YrQMKdOghLhpAPK-BC7Podc2VoHrdfCOcAeKran85ez4V5WXjMJj-Msat3kp--AXsX1q6jzLhKHH7myRLhGfA6iGiuItDHDv_78VlK&wd=&eqid=f218dee7000640da0000000264687cbd)

**Comment by yebangyu** [comment]
thanks for your reply, Edward 

according to the sgd formula:

x_cur = x_prev - learning_rate * grad

if grad is close to zero , we can get that x_cur  is approximately equal to x_prev but not vice versa 

x_cur  is approximately equal to x_prev does not  mean that grad is close to zero (maybe just because learning rate is too small)

Am i right ?
